provider/oracle/environ.go

```diff
@@ -0,0 +1,389 @@
+				Label:     o.cfg.Name(),
+				SSHKeys:   []string{nameKey},
+				Hostname:  o.cfg.Name(),
+				Tags:      []string{o.cfg.Name()},
+				// TODO(sgiulitti): add here the userdata
+				Attributes:  nil,
+				Reverse_dns: false,
+				// TODO(sgiulitti): make vm generate a public address
+			},
+		},
+	})
+
+	result := &environs.StartInstanceResult{
+		Instance: instance,
+	}
+
+	return result, nil
+}
+
+// StopInstances shuts down the instances with the specified IDs.
+// Unknown instance IDs are ignored, to enable idempotency.
+func (o oracleEnviron) StopInstances(...instance.Id) error {
+	return nil
+}
+
+// AllInstances returns all instances currently known to the broker.
+func (o oracleEnviron) AllInstances() ([]instance.Instance, error) {
+	resp, err := o.p.client.AllInstances()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	if len(resp.Result) == 0 {
+		return nil, environs.ErrNoInstances
+	}
+
+	all := make([]instance.Instance, 0, len(resp.Result))
+	for _, val := range resp.Result {
+		inst, err := newInstance(&val)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		all = append(all, inst)
+	}
+
+	return all, nil
+}
+
+// MaintainInstance is used to run actions on jujud startup for existing
+// instances. It is currently only used to ensure that LXC hosts have the
+// correct network configuration.
+func (o oracleEnviron) MaintainInstance(args environs.StartInstanceParams) error {
+	return nil
+}
+
+// Config returns the configuration data with which the Environ was created.
+// Note that this is not necessarily current; the canonical location
+// for the configuration data is stored in the state.
+func (o oracleEnviron) Config() *config.Config {
+	return o.cfg
+}
+
+// ConstraintsValidator returns a constraints.Validator instance which
+// is used to validate and merge constraints.
+//
+// Validator defines operations on constraints attributes which are
+// used to ensure a constraints value is valid, as well as being able
+// to handle overridden attributes.
+//
+// This will use the default validator implementation from the constraints package.
+func (o oracleEnviron) ConstraintsValidator() (constraints.Validator, error) {
+	// list of unsupported oracle provider constraints
+	unsupportedConstraints := []string{
+		constraints.Container,
+		constraints.CpuPower,
+		constraints.RootDisk,
+		constraints.Arch,
+		constraints.VirtType,
+		constraints.Spaces,
+	}
+
+	// we choose to use the default validator implementation
+	validator := constraints.NewValidator()
+	// we must feed the validator that the oracle cloud
+	// provider does not support these constraints
+	validator.RegisterUnsupported(unsupportedConstraints)
+
+	return validator, nil
+}
+
+// SetConfig updates the Environ's configuration.
+//
+// Calls to SetConfig do not affect the configuration of
+// values previously obtained from Storage.
+func (o *oracleEnviron) SetConfig(cfg *config.Config) error {
+	var old *config.Config
```

Comment:
all methods may be (and will be) used concurrently, so don't forget to use mutexes